token-swap/program/src/curve/fees.rs

```diff
@@ -61,6 +61,29 @@ pub fn calculate_fee(
     }
 }
 
+fn ceil_div(dividend: u128, divisor: u128) -> Option<u128> {
+    dividend
+        .checked_add(divisor)?
+        .checked_sub(1)?
+        .checked_div(divisor)
+}
+
+fn pre_fee_amount(
+    post_fee_amount: u128,
+    fee_numerator: u128,
+    fee_denominator: u128,
+) -> Option<u128> {
+    if fee_numerator == 0 {
+        Some(post_fee_amount)
+    } else if fee_numerator == fee_denominator || post_fee_amount == 0 {
+        Some(0)
+    } else {
+        let numerator = post_fee_amount.checked_mul(fee_denominator)?;
+        let denominator = fee_denominator.checked_sub(fee_numerator)?;
+        ceil_div(numerator, denominator)
```

Comment:
shouldn't the inverse be `post_fee_amount * fee_denominator / fee_numerator` ? This seems to say `post_fee_amount * fee_denominator / (fee_denominator - fee_numerator)`

Reply:
the term `inverse` may be confusing. Essentially, given the *output*, we want to know what the *input* was. The typical fee calculation is:

```
output = input * (1 - numerator / denominator)
```

so if we only have the output, we need to do:

```
input = output / (1 - numerator / denominator)
```